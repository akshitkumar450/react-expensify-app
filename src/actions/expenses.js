import { v4 as uuidv4 } from 'uuid';
// Add_expense
export const addExpense = ({ des = '', note = '', amount = 0, createdAt = 0 } = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuidv4(),
        des,
        note,
        amount,
        createdAt
    }
})

// remove_expense
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    expense: {
        id
    }
})
// edit_expense
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    expense: {
        id,
        updates
    }
})
