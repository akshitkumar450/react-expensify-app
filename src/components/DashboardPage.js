import React from 'react';

const DashboardPage = () => (
    <div>
        <p> this is dashboard Page</p>
    </div>
)

export default DashboardPage;