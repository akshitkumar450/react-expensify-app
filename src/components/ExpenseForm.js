import React from 'react';

import moment from 'moment'
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';


// for dates
const date = moment()
// console.log(date.format('MMM Do, YYYY'));

class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.onDescriptionChange = this.onDescriptionChange.bind(this)
        this.onNoteChange = this.onNoteChange.bind(this)
        this.onAmountChange = this.onAmountChange.bind(this)
        this.onDateChange = this.onDateChange.bind(this)
        this.onFocusChange = this.onFocusChange.bind(this)
        // default states
        this.state = {
            description: '',
            note: '',
            amount: '',
            createdAt: moment(),
            calendarFocused: false
        }
    }

    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({ description: description }))
    }

    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({ note: note }))
    }

    onAmountChange(e) {
        const amount = e.target.value
        // we have use regular expression to let user to enter amount in specific format
        // 1000656.54 (only this type)
        if (amount.match(/^\d*(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount: amount }))
        }
    }

    onDateChange(createdAt) {
        this.setState(() => ({ createdAt: createdAt }))
    }

    onFocusChange({ focused }) {
        this.setState(() => ({ calendarFocused: focused }))
    }

    render() {
        return (
            <div>
                <form>
                    <input
                        type='text'
                        placeholder='description'
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />

                    <input
                        type='text'
                        placeholder='amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />

                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />

                    <textarea
                        value={this.state.note}
                        onChange={this.onNoteChange}
                        placeholder='add a note(optional)'
                    ></textarea>
                    <button>add expense</button>
                </form>
            </div>
        )
    }
}
export default ExpenseForm