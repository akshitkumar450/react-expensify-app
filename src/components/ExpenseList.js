import React from 'react'
import { connect } from 'react-redux'

import ExpenseListItem from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'

// step -2
// connect method is used to use the store (connect the component with store)
// using connect method we can use redux store without passing them as the props


// here item is an object of expenses,,so we can pass it as a props 
export const ExpenseList = (props) => (
    <div>
        <h1> Expense List </h1>
        {
            props.expenses.length === 0 ? (
                <p>no expense</p>
            ) : (
                props.expenses.map((item) => {
                    return <ExpenseListItem key={item.id} expense={item} />
                })
            )
        }


    </div>
)

// ALTERNATE TO DO

// const ExpenseList = (props) => (
//     <div>
//         <h1> Expense List </h1>
//         {props.expenses.map((item) => {
//             return <ExpenseListItem key={item.id} {...item} />
//         })}
//     </div>
// )

//  state from the store
//  it has expenses and filters object
// this component can access from store and send to the component with props
//  as the store data changes it will rerender it 

const mapStateToProps = (state) => {
    return {
        // to get the filtered data back
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}

// first arg -> data which we want to use from store
// second ard -> component which we want to use

export default connect(mapStateToProps)(ExpenseList)


// const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList)
// export default ConnectedExpenseList;