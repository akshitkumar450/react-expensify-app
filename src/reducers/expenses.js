// default state for expenses
const expenseReducerDefaultState = []

// all the  action generators will be dispatched to both  reducers
// expense reducer
const expenseReducer = (state = expenseReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // to add a new values to the array
            return [...state, action.expense]

        case 'REMOVE_EXPENSE':
            return state.filter((item) => item.id !== action.expense.id)

        case "EDIT_EXPENSE":
            return state.map((item) => {
                if (item.id === action.expense.id) {
                    return {
                        // it returns a new object with ->(using spread operator)
                        //  to get the old object
                        ...item,
                        //  and to update the old value with the new given value
                        ...action.expense.updates
                    }
                } else {
                    return item
                }
            })

        default:
            return state
    }
}

export default expenseReducer