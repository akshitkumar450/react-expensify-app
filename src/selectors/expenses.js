import moment from 'moment'

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        //  if startDate is undefined then it will return true ( not have effect)
        const createdAtMoment = moment(expense.createdAt)
        // for filtering by date
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true
        // to filter by text
        const textMatch = expense.des.toLowerCase().includes(text.toLowerCase());

        // if all three are true then we will put in the array else it will be filtered out
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        }

        if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
    })

    //the output of the return is an array.so we can use sort on it directly
}

export default getVisibleExpenses