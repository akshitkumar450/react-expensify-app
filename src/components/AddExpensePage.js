import React from 'react';
import ExpenseForm from './ExpenseForm'
import { connect } from 'react-redux'
import { startAddExpense } from '../actions/expenses';

// export class AddExpensePage extends React.Component {
//     constructor(props) {
//         super(props)
//         this.onSubmit = this.onSubmit.bind(this)
//     }

//     onSubmit(expense) {
//         this.props.onSubmit(expense)
//         this.props.history.push('/')
//     }

//     render() {
//         return (
//             <div>
//                 <h1> Add expense</h1>
//                 <ExpenseForm
//                     onSubmit={this.onSubmit}
//                 />
//             </div>
//         )
//     }
// }

const AddExpensePage = (props) => (
    <div>
        <h1> Add expense</h1>
        <ExpenseForm
            // when the form get submitted all data will be sent bt ExpenseForm
            onSubmit={(expense) => {
                // console.log(expense);
                // to add new expenses to store
                props.dispatch(startAddExpense(expense))
                // for redirecting to other page (dashboard)
                // props.onSubmit(expense)
                props.history.push('/')
            }}
        />
    </div>
)
// for testing 
// const mapDispatchToProps = (dispatch) => {
//     //  used to return props
//     return {
//         onSubmit: (expense) => {
//             return dispatch(addExpense(expense))
//         }
//     }
// }
export default connect()(AddExpensePage)
// export default connect(undefined, mapDispatchToProps)(AddExpensePage)