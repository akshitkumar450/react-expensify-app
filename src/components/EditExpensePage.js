import React from 'react';
import { connect } from 'react-redux'
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm'
export class EditExpensePage extends React.Component {
    constructor(props) {
        super(props)
        this.onSubmit = this.onSubmit.bind(this)
        this.onRemove = this.onRemove.bind(this)
    }

    onSubmit(expense) {
        // console.log('updates', expense);
        this.props.startEditExpense(this.props.expense.id, expense)
        this.props.history.push('/')
    }

    onRemove() {
        this.props.startRemoveExpense({ id: this.props.expense.id })
        this.props.history.push('/')
    }

    render() {
        return (
            <div>
                <EditExpensePage
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onRemove}>remove</button>
            </div>
        )
    }

}

// const EditExpensePage = (props) => (
//     <div>
//         <p>id with {props.match.params.id}</p>
//         <ExpenseForm
//             // props.expense will the matched object containing(des,amount,note,createdAt) in expenses array
//             expense={props.expense}
//             onSubmit={(expense) => {
//                 // console.log('updates', expense);
//                 props.dispatch(editExpense(props.expense.id, expense))
//                 props.history.push('/')
//             }} />

//         <button onClick={(e) => {
//             props.dispatch(removeExpense({ id: props.expense.id }))
//             props.history.push('/')
//         }}>remove</button>
//     </div>
// )

const mapStateToProps = (state, props) => {
    return {
        // expense will be the object with the matched id
        // item will the expense object
        expense: state.expenses.find((item) => {
            // to find by id in the expenses array
            return item.id === props.match.params.id
        })
    }
}

const mapDispatchToProps = (dispatch, props) => {
    return {
        startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
        startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)