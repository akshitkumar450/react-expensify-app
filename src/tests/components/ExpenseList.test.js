import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'
// it it the test data
import expenses from '../fixtures/expenses'

test('should render expense list with expense', () => {
    // ExpenseList take expenses array as a props
    const wrapper = shallow(<ExpenseList expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render expense list with  no expense', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})

