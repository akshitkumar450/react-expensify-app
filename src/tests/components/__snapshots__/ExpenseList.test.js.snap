// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list with  no expense 1`] = `
<div>
  <h1>
     Expense List 
  </h1>
  <p>
    no expense
  </p>
</div>
`;

exports[`should render expense list with expense 1`] = `
<div>
  <h1>
     Expense List 
  </h1>
  <ExpenseListItem
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "des": "Gum",
        "id": "1",
        "note": "",
      }
    }
    key="1"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 109500,
        "createdAt": -345600000,
        "des": "Rent",
        "id": "2",
        "note": "",
      }
    }
    key="2"
  />
  <ExpenseListItem
    expense={
      Object {
        "amount": 4500,
        "createdAt": 345600000,
        "des": "Credit Card",
        "id": "3",
        "note": "",
      }
    }
    key="3"
  />
</div>
`;
